dart/simulation/World.cpp

```diff
@@ -87,7 +87,10 @@ World::World(const std::string& _name)
     mPenetrationCorrectionEnabled(false),
     mWrtMass(std::make_shared<neural::WithRespectToMass>()),
     mUseFDOverride(false),
-    mSlowDebugResultsAgainstFD(false)
+    mSlowDebugResultsAgainstFD(false),
+    mConstraintEngine([this](simulation::World* world, bool _resetCommand) {

```

Comment:
Ok, I know this is obnoxious, but I think we need to have much more explicit inputs to the constraint engine callback. Rather than pass a `world` object, I think we need to pass a bunch of `Eigen::MatrixXs` and `Eigen::VectorXs` objects (or vectors of those objects). We also need the constraint engine to return something, not just have side effects on the `world` object.

Basically, this needs to be a pure function with no side effects. Otherwise, when it comes time to differentiate, this is going to be a nightmare to do.

Reply:
To elaborate on that a bit, just know that for every <input, output> pair, if there are multiple inputs and outputs, we'll need to compute a Jacobian relating changes in the input to changes in the output. If there are side effects (hidden inputs and outputs) then that gets a lot harder to detect and do.